### PR TITLE
no global mp lock

### PR DESCRIPTION
**Pull request by thewtex**
- BUG: Do not create global multiprocessing RLock

This addresses:

> /opt/conda/lib/python3.6/multiprocessing/semaphore_tracker.py:143: UserWarning: semaphore_tracker: There appear to be 1 leaked semaphores to clean up at shutdown len(cache))

Reported in Project-MONAI/MONAI#1629

Here we adopt the approach in tqdm:

  tqdm/tqdm#617

to address the similar issues with PyTorch:

  tqdm/tqdm#611

but for our use case. A global set of locks is added to the class on
first use.

We also add a threading lock in addition to a multiprocessing
lock.

### CODE REVIEW

**Comment by blowekamp** [comment]
I'm not sure what to do about the problems caused by mixing process forks and threads that seems to be common in python. Having thread locks indexed by the pid maybe more robust to prevent deadlocks when forking and a thread lock is held.

https://stackoverflow.com/questions/39890363/what-happens-when-a-thread-forks

**Comment by thewtex** [comment]
@blowekamp good point. Since we do not need the thread lock, just the multiprocessing lock to address the pytorch / MONAI use case, I am commenting out the thread lock for now with a note to also investigate a per-pid thread lock.

File: Wrapping/Generators/Python/itk/support/lazy.py [hunk 1]
```diff
@@ -45,6 +33,68 @@ def _lazy_itk_module_reconstructor(module_name, state):
 
     return lazy_module
 
+def _ThreadingRLock(*args, **kwargs):
```
Comment (dzenanz) [inline Wrapping/Generators/Python/itk/support/lazy.py:36]: Perhaps comment out the function, as its invocation is commented out at the moment?

**Comment by dzenanz** [comment]
I modified a MONAI tutorial example to use ITKReader:
https://github.com/dzenanz/MONAI-tutorials/commit/c7a288007462423a8c1b6c1305111a5c82fffeaa
and it works without crashing both before and after the patch.

**Review by dzenanz — APPROVED** [review]

**Comment by thewtex** [comment]
> Since we do not need the thread lock,

CI tells us that this is not true on macOS:

- https://open.cdash.org/viewTest.php?onlyfailed&buildid=7246416
- https://open.cdash.org/viewTest.php?onlyfailed&buildid=7246771

Since the test verifies the use of ThreadPool and tqdm has battle-tested this implementation, let's go with it.

**Pull request merged**